packages/react-dom/src/__tests__/quoteAttributeValueForBrowser-test.js

```diff
@@ -9,40 +9,67 @@
 
 'use strict';
 
+var React;
+var ReactDOMServer;
+
 describe('quoteAttributeValueForBrowser', () => {
-  // TODO: can we express this test with only public API?
-  var quoteAttributeValueForBrowser = require('../shared/quoteAttributeValueForBrowser')
-    .default;
+  beforeEach(() => {
+    jest.resetModules();
+    React = require('react');
+    ReactDOMServer = require('react-dom/server');
+  });
 
-  it('should escape boolean to string', () => {
-    expect(quoteAttributeValueForBrowser(true)).toBe('"true"');
-    expect(quoteAttributeValueForBrowser(false)).toBe('"false"');
+  it('ampersand is escaped inside attributes', () => {
+    var response = ReactDOMServer.renderToString(<img data-attr="&" />);
+    expect(response).toMatch('<img data-attr="&amp;" data-reactroot=""/>');
   });
 
-  it('should escape object to string', () => {
-    var escaped = quoteAttributeValueForBrowser({
-      toString: function() {
-        return 'ponys';
-      },
-    });
+  it('double quote is escaped inside attributes', () => {
+    var response = ReactDOMServer.renderToString(<img data-attr={'"'} />);
+    expect(response).toMatch('<img data-attr="&quot;" data-reactroot=""/>');
+  });
+
+  it('single quote is escaped inside attributes', () => {
+    var response = ReactDOMServer.renderToString(<img data-attr="'" />);
+    expect(response).toMatch('<img data-attr="&#x27;" data-reactroot=""/>');
+  });
 
-    expect(escaped).toBe('"ponys"');
+  it('greater than entity is escaped inside attributes', () => {
+    var response = ReactDOMServer.renderToString(<img data-attr=">" />);
+    expect(response).toMatch('<img data-attr="&gt;" data-reactroot=""/>');
   });
 
-  it('should escape number to string', () => {
-    expect(quoteAttributeValueForBrowser(42)).toBe('"42"');
+  it('lower than entity is escaped inside attributes', () => {
+    var response = ReactDOMServer.renderToString(<img data-attr="<" />);
+    expect(response).toMatch('<img data-attr="&lt;" data-reactroot=""/>');
   });
 
-  it('should escape string', () => {
-    var escaped = quoteAttributeValueForBrowser(
-      '<script type=\'\' src=""></script>',
+  it('number is escaped to string inside attributes', () => {
+    var response = ReactDOMServer.renderToString(<img data-attr={42} />);
+    expect(response).toMatch('<img data-attr="42" data-reactroot=""/>');
+  });
+
+  it('object is passed to a string inside attributes', () => {
+    var sampleObject = {
+      toString: function() {
+        return 'ponys';
+      },
+    };
+
+    var response = ReactDOMServer.renderToString(
+      <img data-attr={sampleObject} />,
     );
-    expect(escaped).not.toContain('<');
-    expect(escaped).not.toContain('>');
-    expect(escaped).not.toContain("'");
-    expect(escaped.substr(1, -1)).not.toContain('"');
+    expect(response).toMatch('<img data-attr="ponys" data-reactroot=""/>');
+  });
 
-    escaped = quoteAttributeValueForBrowser('&');
-    expect(escaped).toBe('"&amp;"');
```

Comment:
So, the nice thing about these old tests is they didn't assert *how* we escape, only that we *do* escape. Which opens the door for changing that in the future.

Maybe it's not too important though.

Reply:
What I didn't like form this test is that thy are just testing a `switch` JavaScript statement by doing it directly from the module. These internal modules pass thourgh other checks and algorithms when applied, so someone could modify that code, prevent the script to actually escape the characters and test would still pass, something that won't happen now 😃

Reply:
Maybe an ideal test would actually create a DOM node, put content into `innerHTML` and then assert that it only has a text node child with expected content?

Reply:
Yeah, that could be one way. Though werid since this is used only on the server package, maybe adding cases from `renderToStaticMarkup` would be valuable.

Reply:
>Though werid since this is used only on the server package

Not sure what you mean. The end goal of rendering to a string is that the HTML shows up in somebody's browser :-) This is exactly what an integration test should verify: that what shows up in the browser is consistent with what we expect (as opposed to a dangerous script tag).

Reply:
I understood it better after writing my response but didn't want to edit it.

Yeah I guess that would mimic the text string from server to node conversion better than just testing the string. Let me know if you are planning on migrating to that. I might think about revisiting these tests in the future if necessary.

Reply:
If you'd like to send a PR for this, happy to review. I'm not 100% sure it's the right approach but it seems to make sense to me.